code/game/mecha/equipment/tools/janitor_tools.dm

```diff
@@ -0,0 +1,363 @@
+	var/output = ..()
+	if(output)
+		return "[output] \[<a href='byond://?src=[UID()];toggle_bagging=1'>[bagging? "Filling" : "Dumping"]</a>\] \[<a href='byond://?src=[UID()];toggle_extended=1'>Area [extended? "Extended" : "Focused"]</a>\] \[Cargo: [length(cargo)]/[cargo_max_weight]</a>\]\]"
+
+/obj/item/mecha_parts/mecha_equipment/janitor/garbage_magnet/Topic(href,href_list)
+	..()
+	var/datum/topic_input/afilter = new (href,href_list)
+	if(afilter.get("toggle_bagging"))
+		bagging = !bagging
+		update_equip_info()
+		return
+	if(afilter.get("toggle_extended"))
+		extended = !extended
+		update_equip_info()
+		return
+
+/obj/item/mecha_parts/mecha_equipment/janitor/garbage_magnet/action(atom/target)
+	if(istype(target, /obj/machinery/disposal)) // Emptying stuff into disposals
+		chassis.occupant.visible_message(
+			"<span class='notice'>[chassis.occupant] empties [src] into the disposal unit.</span>",
+			"<span class='notice'>You empty [src] into disposal unit.</span>",
+			"<span class='notice'>You hear someone emptying something into a disposal unit.</span>"
+		)
+		for(var/obj/item/I in cargo)
+			I.forceMove(target)
+			cargo.Cut()
+		return
+	var/turf/target_turf
+	if(iswallturf(target))
+		return
+	if(isturf(target))
+		target_turf = target
+	else
+		target_turf = target.loc
+	if(bagging) // If picking up
+		if(extended) // If extended reach
+			for(var/turf/current_target_turf in view(1, target_turf))
+				for(var/obj/item/I in current_target_turf.contents)
+					if(can_be_inserted(I))
+						cargo += I
+						I.forceMove(chassis)
+		else // Single turf
+			for(var/obj/item/I in target_turf.contents)
+				if(can_be_inserted(I))
+					cargo += I
+					I.forceMove(chassis)
+		to_chat(chassis.occupant, "<span class='notice'>You pick up all the items with [src]. Cargo compartment capacity: [cargo_max_weight - length(cargo)]</span>")
+	else // Dumping
+		for(var/obj/item/I in cargo)
+			I.forceMove(target_turf)
+			cargo.Cut()
+		to_chat(chassis.occupant, "<span class='notice'>You dump everything out of [src].</span>")
+	update_equip_info()
+
+/obj/item/mecha_parts/mecha_equipment/janitor/garbage_magnet/proc/can_be_inserted(obj/item/I, stop_messages = FALSE)
+	if(!istype(I) || (I.flags & ABSTRACT)) // Not an item
+		return
+
+	if(loc == I)
+		return FALSE // Means the item is already in the storage item
+
+	if(!I.can_enter_storage(src, usr))
+		return FALSE
+
+	if(length(contents) >= cargo_slots)
+		if(!stop_messages)
+			to_chat(chassis.occupant, "<span class='warning'>[I] won't fit in [src], make some space!</span>")
+		return FALSE // Storage item is full
+
+	if(is_type_in_typecache(I, cant_hold)) // Check for specific items which this container can't hold.
+		if(!stop_messages)
+			to_chat(chassis.occupant, "<span class='warning'>[src] cannot hold [I].</span>")
+		return FALSE
+
+	if(length(cant_hold) && isstorage(I)) // Checks nested storage contents for restricted objects, we don't want people sneaking the NAD in via boxes now, do we?
```

Comment:
I think there's a better proc for this that checks for items in contents recursively. After all, a cheese wheel or a book could hold the NAD, not just a storage item.

Also, aren't storage items explicitly blacklisted in cant hold? This feels like a repetitive check.

Reply:
Bags are blacklisted, items like boxes and bibles are not. The check makes sure to check their contents.

Reply:
According to DGamer on Discord, return_inv() is the proc that recursively checks.